### DISCUSSION TITLE
ImageMagick keeps crashing my server (error 503)

### DISCUSSION THREAD

**Discussion by itsthex** [ImageMagick/ImageMagick#5177] [Help]
Hey guys, I'm on a shared server with Ionos (unforutnately). They don't install the ImageMagick module, instead only allowing the command line executable to be used. I am using it in conjunction with Wordpress. 

Long story short, I have a gallery page and it uses very large images, resampling them down to a usable size as well as many thumbnails, per usual form Wordpress. 

The issue is that while I am doing heavy image processing (ie: regenerating thumbnails of 100 images or so, about 5000px each), or even trying to work with the smaller sizes (3000px), after a certain amount of time the server will 503 and leave me dead in the water. It takes down the website completely and I have to call my host (since its shared) to have them stop the background process that is apparently still (unsucessfully) running over and over while the site is 503'd. 

Only at that point will the website come back and I can attempt again, only to have it repeat.

I have read that ImageMagick operates outside the bounds of the PHP memory limits, though I'm not sure if that's true.

I've also tried to set limits in policy.xml, but I have zero idea what to set them to. 

Eventually, I gave up and installed a "clone" copy of my wp on my personal machine and ran everything through localhost after getting imagemagick configured as the php module instead. So far, I am able to do whatever I need as far as batch image processing here. As you can imagine, this is a pain since wordpress also stores metadata in the database for each image, so not only do i have to uplaod thousands of images depending on gallery size, I also have to manually import tabular data for the new images to get them in wordpress.

If anyone could help, I would sincerely appreciate it. I've lost days worth of work to this issue and it's driving me mad, and searching is not providing any solutions that have helped yet. 

Thank you!

**Comment by snibgo** [comment]
I have no experience of Wordpress or running IM on a server.

It would help if you could show the IM commands that are running.

For example, perhaps your commands attempt to read "100 images or so, about 5000px each" into memory at the same time. And perhaps not enough physical memory is available, so IM may eat all the memory and cache other pixels to disk, causing massive loss to performance as well as overwhelming the server.

If that is the case, then re-writing the commands to not read all the images at the same time is the obvious solution. As temporary a test and solution, you could try "-limit disk 0". Then pixels will never be cached to disk. Instead, the command would fail, but that would be better than killing the server.

I suggest also limiting width and height to (for example) 5000 pixels, and time to maybe 30 seconds. Also consider limiting listlength to 10, and memory and map to 1GB.

On limits in general: I suggest setting them so IM can do the work you need it to do, but IM commands will fail if the workload is larger than you expect.

On limits, see https://www.imagemagick.org/script/architecture.php

**Reply by itsthex** [reply]
I wish I could expand on this for you, but I did not write the plugins so I'm not 100% sure how they work. My apologies. This is also the situation with policy.xml  and setting limits. I just don't have a good understanding of those.

**Comment by urban-warrior** [comment]
Adapt the security policy found here: https://imagemagick.org/script/security-policy.php.  We've used this policy for our online image service for tens of years without experiencing any burden to the host.

**Reply by itsthex** [reply]
Thank you for this! I read in this description that there is a limit on the amount of pixels that are written to disk.. is that what the "map" value is?

**Comment by urban-warrior** [comment]
Map is the limit for memory mapping a pixel cache on disk.  The disk limit is the maximum limit of the number of pixels on disk.
